Inital pass at adding home assistant in.

sytone opened the pull request:
Started validation, do not have a card to test moving the data directory so not sure on the best solution there. Please comment on the initial approach. I have hard coded the account and usage of a virtual env as I do not see an example pf prompting the user. The defaults are good however and follow the home assistant approach. 

This is related to #935

sytone committed:
Inital pass at adding home assistant in.

Fourdee commented on dietpi/dietpi-software:
```diff
@@ -11165,6 +11350,30 @@ _EOF_
+
+			# TODO: Validate what happens when user data moves, need to find a SD card. 
+			#mkdir -p "$FP_DIETPI_USERDATA_DIRECTORY"/homeassistant
+			#ln -sf "$FP_DIETPI_USERDATA_DIRECTORY"/homeassistant /home/homeassistant/.homeassistant
```
You may need to also set permissions to the DietPi user data directory, for ```homeassistant``` user.

All permissions are applied here, at the end of installation:
https://github.com/Fourdee/DietPi/blob/master/dietpi/dietpi-software#L7319-L7353

eg:
```
chown -R homeassistant:homeassistant "$FP_DIETPI_USERDATA_DIRECTORY"/homeassistant
```

sytone replied:
must have missed that as I thought I already added it. Whoops! I have updated the apply section now.

Fourdee commented:
@sytone 

Great work, really impressive 👍 

Love the addition of comment notes regarding steps required.

Also need uninstall code for this software title.

Fourdee commented on dietpi/dietpi-software:
```diff
@@ -11165,6 +11350,30 @@ _EOF_
+_EOF_
+
+			# TODO: Validate what happens when user data moves, need to find a SD card. 
+			#mkdir -p "$FP_DIETPI_USERDATA_DIRECTORY"/homeassistant
```
```#mkdir -p "$FP_DIETPI_USERDATA_DIRECTORY"/homeassistant``` not required. The symlink command below will generate this folder.

Fourdee commented:
@sytone 

> do not have a card to test moving the data directory so not sure on the best solution there.

Don't need one. The User data directory exists by default, even on SD card only installs.

If the ```/home/homeassistant/.homeassistant``` directory does not contain any data, then you can just remove the folder and replace with symlink:
```
root@DietPi:~# FP_DIETPI_USERDATA_DIRECTORY='/mnt/dietpi_userdata'

root@DietPi:~# rm -R /home/homeassistant/.homeassistant

root@DietPi:~# ln -sf "$FP_DIETPI_USERDATA_DIRECTORY"/homeassistant /home/homeassistant/.homeassistant

root@DietPi:~# ls -lha /home/homeassistant/
total 8.0K
drwxr-xr-x 2 root root 4.0K Jul 19 13:11 .
drwxr-xr-x 4 root root 4.0K Jul 19 13:07 ..
lrwxrwxrwx 1 root root   34 Jul 19 13:11 .homeassistant -> /mnt/dietpi_userdata/homeassistant

```

sytone committed:
Modifications from review
Update ot the removal script
Addtion of the permissions application

sytone committed:
upstream merge for testing

sytone commented:
The folder will not have anything until the service starts and then it generates its own configuration if there is none there.

Fourdee commented on dietpi/dietpi-software:
```diff
@@ -12096,6 +12309,15 @@ _EOF_
+
+			# How is usage of these items tracked across software packages, what if another package installs
+			# a item in this list and uninstalling HA kills it with this removal?
+			AGP python3 python3-pip libssl-dev cmake libc-ares-dev uuid-dev daemon curl libgnutls28-dev libgnutlsxx28 nmap net-tools sudo libglib2.0-dev cython3 libudev-dev python3-sphinx python3-setuptools libmysqlclient-dev swig libssl-dev python-dev libusb-1.0-0 gcc python3-dev libssl-dev libffi-dev
```
During uninstalls, do not remove shared libraries (eg: python, cmake,  etc) and this could break other software installations. Removing ```-dev``` packages should also be avoided as this may remove something the user needed for their "manual" software installs/projects.

I generally only remove packages/software, if it would "never" effect other software, or the users "general use". Ideally, target the software title, minus possible shared items.

I believe the whole ```AGP``` line could be removed here to be safe.

sytone replied:
ok, thanks for the comment. Ill remove this in the next iteration. Most of the actual install happens in python.

Fourdee commented on dietpi/dietpi-software:
```diff
@@ -337,6 +337,69 @@ _EOF_
 	declare -A aSOFTWARE_AVAIL_HW_MODEL
 	declare -A aSOFTWARE_AVAIL_HW_ARCH
 
```
We need to increase the following item to ```158```, to ensure your new array exists.
```
TOTAL_SOFTWARE_INDEXS_HARDLIMIT=157
```

sytone replied:
I just looked and it is 158. Am I missing something?

Fourdee commented on dietpi/dietpi-software:
```diff
@@ -7343,6 +7464,33 @@ _EOF_
+			su -s /bin/bash homeassistant <<EOF
+virtualenv -p python3 /srv/homeassistant
+EOF
+			su -s /bin/bash homeassistant <<EOF
```
Nice 👍

sytone committed:
Resolved issues with file permissions
- HA now installs and comes up, fture work needed to help handle the update process.
- Added 'hass' to the process tool.
- General commenting.

sytone commented:
ok, I have resolved out some issues with the install process and fixed the access to the configuration. Need tpo look at update process in the future.

Fourdee reviewed (changes requested):
@sytone 

Great work, thanks for making those changes 👍 

We are missing one change, regarding the ```AGP``` removal of shared libs. Please remove this line from the code, then I can merge.

sytone committed:
Removal of AGP line to remove apt packages.

Fourdee commented:
@sytone 

Thanks for the changes, great work 👍

elRadix commented:
@Fourdee  how can I exclude home-assistant from dietpi-service stop/start every time I perform install/update/backup actions? I don't want my home automation to be unavailable during these times.

Fourdee commented:
@elRadix 

Currently, you'd need to modify the source code on your install:
```
/DietPi/dietpi/dietpi-services
```

Remove the line:
https://github.com/Fourdee/DietPi/blob/master/dietpi/dietpi-services#L103

Updating DietPi will however revert your changes, all scripts are overwritten with latest versions.

Ideally we need a "User service exclude file" system/support.

elRadix commented:
@Fourdee thank you

An exclude file would be great indeed same as the include that's already existing.

sytone commented:
IS there a way to add to the existing menus software specific configuration menus? Could a Software Configuration menu be added then apps can register scripts that can configure them? That way HA could have one to enable/disable service restart with the rest of dietpi services. 

Software Optimized   Select DietPi optimized software for installation
Software Additional  Select additional Linux software for installation 
Software Configuration Configure installed software <- This would be the new entry. 

Thoughts?

Fourdee commented:
@sytone 

> Software Configuration Configure installed software <- This would be the new entry.

Love this idea. We could have sections for each peice of installed software, eg:
- List config file locations, selecting them opens it up in nano to edit

> That way HA could have one to enable/disable service restart with the rest of dietpi services.

dietpi-services would need to be modified to read from a "exclude file" to skip those services. We could then generate the exclude file as needed.

sytone commented:
ok, I have been looking at this and how not to duplicate information around the system to make configuration easier. At the moment the configuration is scattered across files. It would be good to try and bring it together. 

- Move to something like python to handle the configuration and scripts in a more controlled way. 
  - PRO: We have all or pythons pre-existing commands available by default. 
  - PRO: Huge amount of modules available to enrich the system
  - CON: Overhead of needing python on all boards 
  - CON: Size considerations. 

- Use jq in bash to move configuration to a json file and out of scripts
  - PRO: All configuration goes to a single json file. 
  - CON: rewrite all scripts to use jq 
  - CON: jq needed on all boards
  - CON: Still in bash and manipulation is not straight forward.

Fourdee commented:
@sytone 

I personally dislike Python, the code layout and syntax simply looks "wrong" to me. Its also inefficient. If we were to move DietPi to another language, I'd personally go with C++ and use a SQL plugin.
It would take months of uninterrupted work to complete this, and thats not going to happen any time soon unfortunately.

Another option, which we use already, is to store everything in a file, then read it like a database, eg:
https://github.com/Fourdee/DietPi/blob/master/dietpi/dietpi-process_tool#L367-L392

Above is awk $X based, however, it could be done on lines, and or, sections using a word start and end trigger based on install index, eg:
```
_START_128
/etc/mpd.conf #mpd conf
/etc/mpd_somethingelse.conf #mpd extra info
_END_128

```

sytone commented:
@Fourdee are you opposed to using something like jq to format the data in json blocks so we are not at the mercy of a type or newline issue for schema problems? This still allows us to clean out the scripts so all the software metadata is in a single place and settings or values can be stored in a local json file alongside the catalogue of software in place. Make a lib script that handles this interaction and just source it in all the files so system config is in one place and user configuration is also preserved between updates. 

Also completely random through, to reduce footprint and still allow separation of the install/uninstall/update/configure scripts for software zip them all up and have the system extract the ones needed for the installed software. Just thinking about maintenance and structure to make it easier for others to contribute. Taking this approach resolves the issue you raised around lots of small files taking up blocks, the zip file would be smaller than the current scripts as well. Yes there is a overhead when pulling the file out for the software package but that should be minimal and items like the conf and misc folder which have software bits in them could be included in this make the file count footprint smaller. For this to work, make a generic library for the zip interaction to get files as needed for execution and then just source it into any script that need to interact with 'non-core' files.

Something like this
```
test.json file contents
{
    "version":"v155",
    "software": [
        {
            "id": 23,
            "name":"LXDE",
            "description":"ultra lightweight desktop",
            "software_category":0,
            "software_type":0,
            "requires":["alsa","xserverxorg"],
            "forum_link":"f=8&t=5#p42",
            "hardware_arch":"",
            "hardware_model":""
        },
        {
            "id": 24,
            "name":"MATE",
            "description":"desktop enviroment",
            "software_category":0,
            "software_type":0,
            "requires":["alsa","xserverxorg"],
            "forum_link":"f=8&t=5&start=60#p2073",
        }        
    ]
}

root@homeassistant:~# cat test.json | jq '.software[] | select(.id == 24)'
{
  "id": 24,
  "name": "MATE",
  "description": "desktop enviroment",
  "software_category": 0,
  "software_type": 0,
  "requires": [
    "alsa",
    "xserverxorg"
  ],
  "forum_link": "f=8&t=5&start=60#p2073"
}
root@homeassistant:~# cat test.json | jq '.software[] | select(.name == "MATE")'
{
  "id": 24,
  "name": "MATE",
  "description": "desktop enviroment",
  "software_category": 0,
  "software_type": 0,
  "requires": [
    "alsa",
    "xserverxorg"
  ],
  "forum_link": "f=8&t=5&start=60#p2073"
}

root@homeassistant:~# cat test.json | jq '.software[].id' | while read object; do
>   echo "$object"
> done
23
24

--




```

Fourdee commented:
@sytone 

>  are you opposed to using something like jq to format the data in json blocks so we are not at the mercy of a type or newline issue for schema problems? 

Not at all, if it cleans up the code and makes the workflow easier for everyone, i'am all for it. 
Ideally, if we could have a working prototype example of how this would work, we can see if it benefits the project contributors and our users?

> Also completely random through, to reduce footprint and still allow separation of the install/uninstall/update/configure scripts for software zip them all up and have the system extract the ones needed for the installed software. Just thinking about maintenance and structure to make it easier for others to contribute.

I really like the idea of having the install/uninstall sections contained within unique script files, but each script file will need to support the global vars we use in ```dietpi-software``` (eg: $HW_MODEL / $HW_ARCH / $FP_DIETPIUSERDATA etc). 
Could try and take this further by globally setting vars at system level, eg: ```export VAR=value``` etc, then we wouldn't have to file load these in every script. But, they would need to be updated/refreshed at times (global var refresh script? But would slow down system unless we define which vars we actually need to update).

Regardless, we shouldn't focus on "over-complicating", instead, "simplifying", as long as it benefits project workflow for contributors and our end users.
But I completely agree that ```dietpi-software``` should be cleaned up to remove the install/uninstall (and array) code, instead having it located externally and imported on demand.

elRadix commented:
Is there a way to upgrade the next version of python without breaking the system?

`[homeassistant.bootstrap] Python 3.4 support has been deprecated and will be removed in the begining of 2018. Please upgrade Python or your operating system. More info: https://home-assistant.io/blog/2017/10/06/deprecating-python-3.4-support/`

Fourdee commented:
@elRadix 

Easiest solution is to upgrade to Stretch:
```
python3 is already the newest version (3.5.3-1)
```

https://github.com/Fourdee/DietPi/issues/475#issue-171250292

elRadix commented:
@Fourdee, I already have a lot of apps running, upgrading is only possible if I install a new image and start from scratch? 

Is there an easy way to upgrade to Stretch or upgrading python?
I also think using diepti-backup to restore my config might not work on Strecht

EDIT: followed this guide to update python and updated virtual env only but adjusting the correct paths
https://community.home-assistant.io/t/python-3-6-upgrade-of-a-virtualenv/21722/87
```

#Download and install latest python
cd /home/pi
wget https://www.python.org/ftp/python/3.6.3/Python-3.6.3.tgz
tar xzvf Python-3.6.3.tgz
cd Python-3.6.3/
./configure
make
sudo make install

Note: Last step takes a while.

#Delete old virtual environment
cd /srv/
sudo rm -r *

#Create new virtual environment with latest python
cd /srv/
sudo mkdir homeassistant
sudo chown homeassistant:homeassistant homeassistant
sudo su -s /bin/bash homeassistant
cd /srv/homeassistant
python3.6.3 -m venv .
source bin/activate

#Install latest Hass
pip3 install homeassistant
hass

Note:  First start takes a while
```

sytone commented:
I have a pull that has the pyenv in it. It allows better control of the python versions and allow multiple versions to be installed and not impact the distribution version.

Fourdee commented:
@sytone 

> I have a pull that has the pyenv in it. It allows better control of the python versions and allow multiple versions to be installed and not impact the distribution version.

Thanks. I've merged this for testing: https://github.com/Fourdee/DietPi/issues/1145#issuecomment-339755566
